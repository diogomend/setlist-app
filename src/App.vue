<template>
  <v-app>
    <v-progress-linear
      v-if="isLoading"
      color="deep-orange accent-4"
      indeterminate
      rounded
      height="6"
    ></v-progress-linear>
    <template v-else-if="getTotalConcerts > 0">
      <app-bar v-if="isMobile" />
      <nav-drawer />
      <v-container class="main-container">
        <router-view />
      </v-container>
    </template>
    <template v-else>
      <h3 class="mt-10">Couldn't load Setlist for your account</h3>
    </template>
  </v-app>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
import AppBar from "@/components/AppBar";
import NavDrawer from "@/components/NavDrawer";
import mobileMixin from "@/mixins/mobileMixin";
export default {
  components: {
    AppBar,
    NavDrawer
  },
  mixins: [mobileMixin],
  methods: {
    ...mapActions("setlist", ["fetchSetlists"])
  },
  computed: {
    ...mapGetters({
      isLoading: "appState/getLoading",
      getTotalConcerts: "setlist/getTotalConcerts"
    })
  },
  mounted() {
    this.fetchSetlists();
  }
};
</script>

<style lang="scss">
body .v-card--link:focus:before {
  opacity: 0;
}
.hide-ripple .v-ripple__container,
.v-ripple__container {
  display: none;
}

.relative {
  position: relative;
}

#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
}

#nav {
  padding: 30px;

  a {
    font-weight: bold;
    color: #2c3e50;

    &.router-link-exact-active {
      color: #42b983;
    }
  }
}

.container.main-container {
  width: 90%;
  max-width: initial;
  padding-top: 100px;

  @media screen and (min-width: 960px) {
    padding-left: 5%;
  }
}
</style>
