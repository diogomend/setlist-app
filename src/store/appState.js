const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  }
};

const mutations = {
  TOGGLE_DRAWER(state, bool) {
    state.drawer = bool;
  },
  TOGGLE_LOADING(state, isLoading) {
    state.isLoading = isLoading;
  }
};

const getters = {
  getDrawer: state => state.drawer,
  getLoading: state => state.isLoading
};

const state = () => ({
  drawer: false,
  isLoading: true
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
