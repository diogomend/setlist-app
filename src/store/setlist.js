import setlistService from "@/service/setlist";

const actions = {
  fetchSetlists: async ({ commit, dispatch }) => {
    try {
      const { data } = await setlistService.getSetlists();
      commit("UPDATE_SETLISTS", data);
      commit("UPDATE_ARTISTS", data);
      //return Promise.resolve(data);
      return true;
    } catch (err) {
      //return Promise.reject('Could not load setlist information');
      return false;
    } finally {
      dispatch("appState/toggleLoading", false, { root: true })
    }
  }
};

const mutations = {
  UPDATE_SETLISTS(state, data) {
    state.setlists = data;
  },
  UPDATE_ARTISTS(state, data) {
    state.artists = getArtists(data);
  }
};

const getters = {
  getTotalConcerts: state => state.setlists.length,
  getShow: state => id => {
    return Object.values(state.setlists).find(element => element.id === id);
  },
  getArtist: state => mbid => {
    const data = Object.values(state.setlists).find(
      element => element.artist.mbid === mbid
    );
    if (data && Object.prototype.hasOwnProperty.call(data, "artist")) {
      return data.artist;
    }

    return null;
  },
  getArtists: state => {
    const artists = getArtists(state.setlists);

    return Object.values(artists).sort((a, b) => (a.name > b.name ? 1 : -1));
  },
  getArtistShows: state => mbid => {
    return state.setlists.reduce((ac, setlist) => {
      if (setlist.artist.mbid === mbid) {
        ac.push(setlist);
      }

      return ac;
    }, []);
  },
  getTopArtists: state => {
    const artists = Object.values(state.artists).sort((a, b) =>
      a.qty < b.qty ? 1 : -1
    );
    return artists.slice(0, 6);
  },
  getTotalArtists: state => Object.keys(state.artists).length,
  getYearlyConcerts: state => {
    return state.setlists.reduce((ac, setlist) => {
      const dateSplit = setlist.eventDate.split("-");
      const year = parseInt(dateSplit[2]);

      if (Object.prototype.hasOwnProperty.call(ac, year)) {
        const merge = [...ac[year], setlist];
        return { ...ac, [year]: merge };
      }

      return { ...ac, [year]: [setlist] };
    }, {});
  },
  getTotalCountries: state => {
    const countries = state.setlists.reduce((ac, setlist) => {
      const country = setlist.venue.city.country.code;
      const qty = Object.prototype.hasOwnProperty.call(ac, country)
        ? ac[country] + 1
        : 1;

      return { ...ac, [country]: qty };
    }, {});

    return Object.keys(countries).length;
  }
};

const state = () => ({
  setlists: [],
  isLoading: true
});

const getArtists = setlists => {
  return setlists.reduce((ac, setlist) => {
    setlist.artist.qty = Object.prototype.hasOwnProperty.call(
      ac,
      setlist.artist.mbid
    )
      ? ac[setlist.artist.mbid].qty + 1
      : 1;

    return { ...ac, [setlist.artist.mbid]: setlist.artist };
  }, {});
};

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
