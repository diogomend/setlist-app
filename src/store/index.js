import Vue from "vue";
import Vuex from "vuex";
import setlist from "@/store/setlist";
import appState from "@/store/appState";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    setlist,
    appState
  }
};

export default new Vuex.Store(storeOptions);
