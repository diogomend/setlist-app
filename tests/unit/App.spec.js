import { createLocalVue, shallowMount } from "@vue/test-utils";
import flushPromises from "flush-promises";
import axios from "axios";
import App from "@/App";
import Vuex from "vuex";
import Vue from "vue";
import Vuetify from "vuetify";
import { storeOptions } from "@/store/index";
Vue.use(Vuetify);
jest.mock("axios");

let localVue = createLocalVue();

describe("App.vue", () => {
  let store = new Vuex.Store(storeOptions);
  it("is a Vue instance", () => {
    const wrapper = shallowMount(App, {
      localVue,
      store,
      sync: false
    });

    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });

  it("Should show message after bad response", async () => {
    //const fetchSetlists = jest.fn().mockReturnValueOnce({});
    axios.get.mockImplementationOnce(() => Promise.reject());
    const wrapper = shallowMount(App, {
      localVue,
      store,
      methods: {
        //fetchSetlists
      }
    });
    await flushPromises();

    const errorContent = wrapper.find("h3");
    expect(errorContent.text()).toBe("Couldn't load Setlist for your account");
  });

  it("Should show progressbar if state is loading", () => {
    const fetchSetlists = jest.fn().mockReturnValueOnce({});
    shallowMount(App, {
      localVue,
      store,
      methods: {
        fetchSetlists
      }
    });

    expect(fetchSetlists).toHaveBeenCalled();
  });
});
