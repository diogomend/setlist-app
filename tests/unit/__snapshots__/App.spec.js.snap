// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue is a Vue instance 1`] = `
<v-app-stub id="app">
  <v-progress-linear-stub color="deep-orange accent-4" value="0" active="true" buffervalue="100" height="6" indeterminate="true" rounded="true"></v-progress-linear-stub>
</v-app-stub>
`;
